### ISSUE TITLE
clipman.init() can't decode byte 0xa0 error.

### ISSUE THREAD

**Issue by pekcheey** [NikitaBeloglazov/clipman#21]
Sometimes I get the following error when calling clipman.init()

```python-traceback
Traceback (most recent call last):
......
  File "/opt/homebrew/Cellar/python@3.12/3.12.6/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/clipman/__init__.py", line 83, in run_command
    return runner.stdout.decode("UTF-8").removesuffix("\n") # looks like all commands returns \n in the end
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xa0 in position 48: invalid start byte
```

from the following 

[†](https://stackoverflow.com/questions/22216076/unicodedecodeerror-utf8-codec-cant-decode-byte-0xa5-in-position-0-invalid-s) | python - UnicodeDecodeError: 'utf8' codec can't decode byte 0xa5 in position 0: invalid start byte - Stack Overflow

the solution is to make a call to strip() before decode()
I tried the following  and it seems to resolve the error. 

```diff
src/clipman/__init__.py
85c85
< 	return runner.stdout.decode("UTF-8").removesuffix("\n") # looks like all commands returns \n in the end
---
> 	return runner.stdout.strip().decode("UTF-8").removesuffix("\n") # looks like all commands returns \n in the end
```

**Comment by pekcheey** [comment]
apparently the above diff doesn't fix the issue. still checking. 

Seems like adding a ', errors="ignore"' to the decode call works more reliably.

```diff
-       return runner.stdout.decode("UTF-8").removesuffix("\n") # looks like all commands returns \n in the end
+       return runner.stdout.decode("UTF-8", errors="ignore").removesuffix("\n") # looks like all commands returns \n in the end

```

**Comment by NikitaBeloglazov** [comment]
Hello!
Please show output of `clipman.init(debug=True)` and `pip3 show clipman`

**Comment by pekcheey** [comment]
ouch.  look like it is probably an one of those issues that cannot be replicated after a computer restart. 

``` bash
> pip3.12 show clipman
Name: clipman
Version: 3.2.3
Summary: Python3 module for working with clipboard. Created because pyperclip is discontinued.
Home-page: https://github.com/NikitaBeloglazov/clipman
Author: 
Author-email: NikitaBeloglazov <nnikita.beloglazov@gmail.com>
License: MPL 2.0 License
Location: /opt/homebrew/lib/python3.12/site-packages
Requires: 
Required-by: 
```

I think we should close this issue for now. I'll be sure to send the output of clipman.init(debug=True) ifI encounter the issue again.

**Comment by NikitaBeloglazov** [comment]
You using old version of clipman
This already get fixed in v3.2.5
Please see issue #16

**Comment by pekcheey** [comment]
oh I see. Okay. I'll update. Thanks for the quick response.

**Comment by pekcheey** [comment]
for what it's worth when I have the following cross like character in the clipboard at the time of calling clipman init, the problem still happen. [†]

after some testing, it seems that forcing an encoding when calling subprocess works.

```python
 runner = subprocess.run(command, timeout=timeout, shell=False, check=False, stdout=subprocess.PIPE, stderr=subprocess.P IPE, encoding="ISO-885-1")  
                                                                                                                            ▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲▲   
```

with the encoding done within subprocess.run, the subsequent call to  `.decode('UTF-8')` is no longer needed.

apparently, using the `encoding="unicode_escape"` works as well.

that said, is this a "fix" or a "workaround" for this particular case? I don't know.
it may depends on factors such as 
  * what is the user copying. 
  * whether it affects an eventual translation back into the more commonly used "UTF-8" (I suspect no)
  * performance penalty if any (I suspect no)

after a bit of googling / asking ChatGPT (info may be sketchy) 
it seems that ISO-8859-1 is a single byte encoding compared to UTF-8 and unicode_escape
I suspect that with regard to Clipman, the encoding selected should be the more permissive encoding (that can encode more character's without throwing error)

Of course, it is also possible that these sort of errors may be regarded as special cases. And I suspect that for the majority of users, this issue is not encountered. and this issue may be close again. I don't mind it. I'm just going to reopen the issue with the comment for your attention. 

Thanks for creating clipman!

**Comment by NikitaBeloglazov** [comment]
**Hello again!** ☺️

Cannot reproduce this on macOS HighSierra.

Please provide additional information about your OS such as:
- `python3 -c "import clipman; clipman.init(debug=True); print(clipman.get())"`
- `pip3 show clipman`
- OS version info
- Also, good to see output of `locale charmap`

_P.S. I've now noticed that you using homebrew... The thing is that I don’t have a MacBook for a testing ground and installing it on virtual machine is not so simple.. I have old macOS HighSierra VM but it's maybe too old for this issue.._

---

**About changing default encoding UTF-8 to ISO-8859-1:**
No, because UTF-8 is a standart in UNIX and potentially can break some things....
I can change the encoding personally for MacOS, but I still doubt that MacOS is now using a different encoding

---

**Buuut, about stop using `.decode('UTF-8')` in favor of `encoding=UTF-8 in subprocess.run`**
That's a definetely good idea, I didn't notice this thing, maybe i'll do this :)

**Comment by pekcheey** [comment]
After some more testing it seems that there is no problem with clipman. The problem is with MacVim.

First, the info you wanted.

I'm usually calling clipman from within the MacVIM editor python integration which is pointing to specifically version 3.13 on my machine

```
> python3.13 -c "import clipman; clipman.init(debug=True); print(clipman.get())"
Init call start
OS family: posix
posix.uname_result(sysname='Darwin', nodename='cayces-mac-mini', release='22.6.0', version='Darwin Kernel Version 22.6.0: Mon Jun 24 01:21:28 PDT 2024; root:xnu-8796.141.3.706.2~1/RELEASE_ARM64_T8103', machine='arm64')
Detected OS: macOS
Detected engine: pboard
python3 -c "import clipman; clipman.init(debug=True); print(clipman.get())"

> pip3.13 show clipman
Name: clipman
Version: 3.4.0.dev3+g9cf39db
Summary: Python3 module for working with clipboard. Created because pyperclip is discontinued.
Home-page: https://github.com/NikitaBeloglazov/clipman
Author: 
Author-email: NikitaBeloglazov <nnikita.beloglazov@gmail.com>
License: MPL 2.0 License
Location: /opt/homebrew/lib/python3.13/site-packages
Requires: dbus-next
Required-by: 

> sw_vers
ProductName:		macOS
ProductVersion:		13.6.9
BuildVersion:		22G830

> sysctl machdep.cpu.brand_string
machdep.cpu.brand_string: Apple M1

> machine
arm64e

> locale charmap
US-ASCII


```

btw, I should mentioned that it seems that python ships with default UTF-8 enoding.

```
> python3.13 -c "import sys; print(sys.getdefaultencoding())"
utf-8
```

The following show that the problem is with MacVim

In Shell
```
> echo [†] | pbcopy
> python3.13 -c "import clipman; clipman.init(debug=True); print(clipman.get())"
Init call start
OS family: posix
posix.uname_result(sysname='Darwin', nodename='cayces-mac-mini', release='22.6.0', version='Darwin Kernel Version 22.6.0: Mon Jun 24 01:21:28 PDT 2024; root:xnu-8796.141.3.706.2~1/RELEASE_ARM64_T8103', machine='arm64')
Detected OS: macOS
Detected engine: pboard
[†]
```
the above command works.

Via MacVim command execution works if copying through pbcopy  (notice the "r!")
```
r! echo [†] | pbcopy
r!python3.13 -c "import clipman; clipman.init(debug=True); print(clipman.get())"
Init call start
OS family: posix
posix.uname_result(sysname='Darwin', nodename='cayces-mac-mini', release='22.6.0', version='Darwin Kernel Version 22.6.0: Mon Jun 24 01:21:28 PDT 2024; root:xnu-8796.141.3.706.2~1/RELEASE_ARM64_T8103', machine='arm64')
Detected OS: macOS
Detected engine: pboard
[†]
```


but if I were to manually copy the string "[†]" in MacVim and execute it directly in MacVim through the command execution
I'll be getting the error.

```
r!python3.13 -c "import clipman; clipman.init(debug=True); print(clipman.get())"
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xa0 in position 1: invalid start byte
Init call start
OS family: posix
posix.uname_result(sysname='Darwin', nodename='cayces-mac-mini', release='22.6.0', version='Darwin Kernel Version 22.6.0: Mon Jun 24 01:21:28 PDT 2024; root:xnu-8796.141.3.706.2~1/RELEASE_ARM64_T8103', machine='arm64')
Detected OS: macOS
Traceback (most recent call last):
  File "<string>", line 1, in <module>
    import clipman; clipman.init(debug=True); print(clipman.get())
                    ~~~~~~~~~~~~^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.13/site-packages/clipman/__init__.py", line 353, in init
    dataclass.engine = detect_clipboard_engine()
                       ~~~~~~~~~~~~~~~~~~~~~~~^^
  File "/opt/homebrew/lib/python3.13/site-packages/clipman/__init__.py", line 238, in detect_clipboard_engine
    return check_run_command(['pbpaste'], "pboard")
  File "/opt/homebrew/lib/python3.13/site-packages/clipman/__init__.py", line 116, in check_run_command
    run_command(command, features=features)
    ~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.13/site-packages/clipman/__init__.py", line 95, in run_command
    parcel =  runner.stdout.decode('UTF-8').removesuffix("\n") # looks like all commands returns \n in the end
              ~~~~~~~~~~~~~~~~~~~~^^^^^^^^^
```

btw, changing the encoding to either `unicode_escape` or `ISO-8859-1` resulted in the output below which is obviously wrong.

```
[ ]
```
so scratch that approach.

In summary, since the problem does not lies with clipman, I'll close this issue again.

**Comment by NikitaBeloglazov** [comment]
Okay, thank you for feedback! 😇
